pootle/static/js/utils.js

```diff
@@ -78,62 +74,111 @@ var strCmp = function (a, b) {
 };
 
 
-/* Cleans '\n' escape sequences and adds '\t' sequences */
-var cleanEscape = function (s) {
-  return s.replace(/\\t/g, "\t").replace(/\\n/g, "");
-};
+const PUNCTUATION_RE = /[™©®]|[℃℉°]|[±πθ×÷−√∞∆Σ′″]|[‘’ʼ‚‛“”„‟]|[«»]|[£¥€]|…|—|–|[ ]/;
+// Marks + Degree-related + Maths + Quote characters + Guillemets + Currencies +
+// U2026 horizontal ellipsis + U2014 em dash + U2013 en dash +
+// U202F narrow no-break space
+
+function highlightPunctuation(text) {
+  function replace(match) {
+    return `<span class="highlight-punctuation js-editor-copytext">${match}</span>`;
+  }
 
+  return text.replace(PUNCTUATION_RE, replace);
+}
 
-/* Fancy escapes to highlight parts of the text such as HTML tags */
-var fancyEscape = function (text) {
+
+const nonPrintable = '<span class="non-printable js-editor-copytext %1$s" ' +
+                     'data-string="%2$s"></span>';
+const newLineHl = nonPrintable.replace('%1$s', 'newline')
+                              .replace('%2$s', '&#10;') + '<br/>';
+const tabHl = nonPrintable.replace('%1$s', 'tab').replace('%2$s', '&#9;');
+const escapeHl = '<span class="highlight-escape js-editor-copytext">%s</span>';
+
+const ESCAPE_RE = /\r\n|[\r\n\t]|\\r|\\n|\\t/gm;
+
+
+function highlightEscapes(text) {
 
   function replace(match) {
-      var replaced,
-          escapeHl= '<span class="highlight-escape">%s</span>',
-          htmlHl = '<span class="highlight-html">&lt;%s&gt;</span>',
-          submap = {
-            '\r\n': escapeHl.replace(/%s/, '\\r\\n') + '<br/>\n',
-            '\r': escapeHl.replace(/%s/, '\\r') + '<br/>\n',
-            '\n': escapeHl.replace(/%s/, '\\n') + '<br/>\n',
-            '\t': escapeHl.replace(/%s/, '\\t'),
-            '&': '&amp;',
-            '<': '&lt;',
-            '>': '&gt;'
-          };
-
-      replaced = submap[match];
-
-      if (replaced === undefined) {
-        replaced = htmlHl.replace(
-            /%s/,
-            fancyEscape(match.slice(1, match.length-1))
-        );
-      }
-
-      return replaced;
+      const submap = {
+        '\r\n': newLineHl,
+        '\r': newLineHl,
+        '\n': newLineHl,
+        '\t': tabHl,
+        '\\r': escapeHl.replace(/%s/, '\\r'),
+        '\\n': escapeHl.replace(/%s/, '\\n'),
+        '\\t': escapeHl.replace(/%s/, '\\t'),
+      };
+
+      return submap[match];
   }
 
-  return text.replace(escapeRE, replace);
-};
+  return text.replace(ESCAPE_RE, replace);
+}
 
 
-/* Highlight spaces to make them easily visible */
-var fancySpaces = function (text) {
+const spaceHl = '<span class="highlight-whitespace js-editor-copytext" data-string="%s"></span>';
+const WHITESPACE_RE = /^([ \u{00a0}]+)|([ \u{00a0}]+)$|[ \u{00a0}]{2,}/gmu;
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/dwayne/pootle/file/2059748190/issues/source?bid=1990595&fileBranchId=2945338#l122)